Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,745 @@
+/-
+Copyright (c) 2025 Silvère Gangloff. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Silvère Gangloff
+-/
+import Mathlib.Topology.Constructions
+
+/-!
+# Symbolic dynamics on cancellative monoids
+
+This file develops a minimal API for symbolic dynamics over an arbitrary
+**right-cancellative monoid** `G` (`[Monoid G] [IsRightCancelMul G]`).
+
+Given a finite alphabet `A`, the ambient configuration space is the set of
+functions `G → A`, endowed with the product topology. We define the
+right-translation action, cylinders, finite patterns, their occurrences,
+forbidden sets, and subshifts (closed, shift-invariant subsets). Basic
+topological facts (e.g. cylinders are clopen, occurrence sets are clopen,
+forbidden sets are closed) are proved under discreteness assumptions on
+the alphabet.
+
+The development is generic for right-cancellative monoids. This covers both
+groups (the standard setting of symbolic dynamics) and more general monoids
+where cancellation holds but inverses may not exist. Geometry specific to
+`ℤ^d` (boxes/cubes and the box-based entropy) is deferred to a separate
+specialization.
+
+## Why cancellativity?
+
+Some constructions, such as translating a finite pattern to occur at a point `v`,
+require solving equations of the form `w * v = h`. For this to have a unique
+solution `w` given `h` and `v`, we assume **right-cancellation**:
+if `a * v = b * v` then `a = b`. This allows us to define
+`patternToConfig` (which extends a pattern into a configuration
+using the default value of `A`) without using inverses,
+so that the theory works not only for groups but also for cancellative monoids.
+
+## Main definitions
+
```

Comment:
Maybe say that you are only going to describe the additive ones, although multiplicative versions also exist for all the notions (and adapt)? Most of what you write below is additive, right?